p2p/syncer.go

```diff
@@ -536,6 +538,34 @@ func (s *Syncer) syncToPeer(peer gateway.Header) {
 	}
 }
 
+// maintainHealthyPeerSet tries to add peers to the syncer until it has 8.
+func (s *Syncer) maintainHealthyPeerSet() {
+	for {
+		s.peerCond.L.Lock()
+		for len(s.peers) >= 8 {
+			// peerCond is signalled upon disconnecting from a peer
+			s.peerCond.Wait()
+		}
+		peer, err := s.store.RandomPeer()
+		s.peerCond.L.Unlock()
+		if err != nil {
+			continue
+		}
+
+		exists := false
+		for _, connected := range s.Peers() {
+			if peer == connected.NetAddress {
+				exists = true
+			}
+		}
```

Comment:
for this, you'll want to create a `seen := make(map[string]bool)` at the top of `maintainHealthyPeerSet`. This will prevent you from repeatedly attempting to connect to unreachable peers.